src/input_common/gcadapter/gc_adapter.cpp

```diff
@@ -0,0 +1,371 @@
+        pad.trigger_right = adapter_payload[1 + (9 * port) + 8];
+    }
+    return pad;
+}
+
+void Adapter::PadToState(const GCPadStatus& pad, GCState& state) {
+    state.buttons.insert_or_assign(PAD_BUTTON_A, pad.button & PAD_BUTTON_A);
+    state.buttons.insert_or_assign(PAD_BUTTON_B, pad.button & PAD_BUTTON_B);
+    state.buttons.insert_or_assign(PAD_BUTTON_X, pad.button & PAD_BUTTON_X);
+    state.buttons.insert_or_assign(PAD_BUTTON_Y, pad.button & PAD_BUTTON_Y);
+    state.buttons.insert_or_assign(PAD_BUTTON_LEFT, pad.button & PAD_BUTTON_LEFT);
+    state.buttons.insert_or_assign(PAD_BUTTON_RIGHT, pad.button & PAD_BUTTON_RIGHT);
+    state.buttons.insert_or_assign(PAD_BUTTON_DOWN, pad.button & PAD_BUTTON_DOWN);
+    state.buttons.insert_or_assign(PAD_BUTTON_UP, pad.button & PAD_BUTTON_UP);
+    state.buttons.insert_or_assign(PAD_BUTTON_START, pad.button & PAD_BUTTON_START);
+    state.buttons.insert_or_assign(PAD_TRIGGER_Z, pad.button & PAD_TRIGGER_Z);
+    state.buttons.insert_or_assign(PAD_TRIGGER_L, pad.button & PAD_TRIGGER_L);
+    state.buttons.insert_or_assign(PAD_TRIGGER_R, pad.button & PAD_TRIGGER_R);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::StickX), pad.stick_x);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::StickY), pad.stick_y);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::SubstickX), pad.substick_x);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::SubstickY), pad.substick_y);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::TriggerLeft), pad.trigger_left);
+    state.axes.insert_or_assign(static_cast<u8>(PadAxes::TriggerRight), pad.trigger_right);
+}
+
+void Adapter::Read() {
+    LOG_INFO(Input, "GC Adapter Read() thread started");
+
+    int payload_size_in, payload_size;
+    std::array<u8, 37> adapter_payload;
+    std::array<u8, 37> controller_payload_copy;
+    std::array<GCPadStatus, 4> pad;
+
+    while (adapter_thread_running) {
+        libusb_interrupt_transfer(usb_adapter_handle, input_endpoint, adapter_payload.data(),
+                                  sizeof(adapter_payload), &payload_size_in, 32);
+        payload_size = 0;
+        {
+            std::lock_guard<std::mutex> lk(s_mutex);
+            std::copy(std::begin(adapter_payload), std::end(adapter_payload),
+                      std::begin(controller_payload_copy));
+            payload_size = payload_size_in;
+        }
+
+        if (payload_size != sizeof(controller_payload_copy) ||
+            controller_payload_copy[0] != LIBUSB_DT_HID) {
+            // TODO: It might be worthwhile to Shutdown GC Adapter if we encounter errors here
+            LOG_ERROR(Input, "error reading payload (size: %d, type: %02x)", payload_size,
+                      controller_payload_copy[0]);
+        } else {
+            for (int port = 0; port < 4; port++) {
+                pad[port] = CheckStatus(port, controller_payload_copy);
+            }
+        }
+        for (int port = 0; port < 4; port++) {
```

Comment:
Is this the size of pad? If so:
```suggestion
        for (int port = 0; port < pad.size(); port++) {
```